blocks/search.py

```diff
@@ -0,0 +1,280 @@
+        Computational graph which contains `sequence_generator`.
+
+    See Also
+    --------
+    :class:`SequenceGenerator`, :class:`SequenceContentAttention`.
+
+    Notes
+    -----
+    Sequence generator should use an emitter which has `probs` method and
+    one of its outputs is called `probs` e.g. :class:`SoftmaxEmitter`.
+
+    """
+    def __init__(self, beam_size, sequence_generator, comp_graph):
+        self.beam_size = beam_size
+        self.sequence_generator = sequence_generator
+        self.generate_names = sequence_generator.generate.states
+        self.context_names = sequence_generator.generate.contexts
+        self.init_computer = None
+        self.next_state_computer = None
+        self.context_computer = None
+        self.initial_state_computer = None
+        self.state_names = (sequence_generator.state_names +
+                            sequence_generator.glimpse_names +
+                            ['outputs'])
+        self.context_names = sequence_generator.context_names
+        self.comp_graph = comp_graph
+        self.inputs = OrderedDict(comp_graph.dict_of_inputs())
+        self.need_input_states = []
+        self.compiled = False
+
+    def compile_context_computer(self, contexts):
+        """Compiles `context_computer`."""
+        self.context_computer = function(list(self.inputs.values()),
+                                         list(contexts.values()),
+                                         on_unused_input='ignore')
+
+    def compile_initial_state_computer(self, generator, contexts):
+        """Compiles `initial_state_computer`."""
+        initial_states = []
+        for name in self.state_names:
+            initial_states.append(generator.initial_state(
+                name,
+                self.beam_size,
+                **contexts))
+        self.initial_state_computer = function(list(contexts.values()),
+                                               initial_states,
+                                               on_unused_input='ignore')
+
+    def compile_next_state_computer(self, generator, contexts, inner_cg):
+        """Compiles `next_state_computer`."""
+        states = []
+        for name in generator.generate.states:
+            var = VariableFilter(bricks=[generator], name='^' + name + '$',
+                                 roles=[INPUT])(inner_cg)[-1:]
+            if var:
+                self.need_input_states.append(name)
+            states.extend(var)
+
+        next_states = []
+        for name in generator.generate.states:
+            var = VariableFilter(bricks=[generator], name='^' + name + '$',
+                                 roles=[OUTPUT])(inner_cg)[-1:]
+            next_states.extend(var)
+
+        next_probs = VariableFilter(
+            bricks=[generator.readout.emitter],
+            name='^probs')(inner_cg)[-1]
+        # Create theano function for next states
+        self.next_state_computer = function(list(contexts.values()) + states,
```

Comment:
The problem is here: https://github.com/bartvm/blocks/blob/master/blocks/bricks/sequence_generators.py#L208, and I tried to explain it in first comment. This line is still used by your code.

You can also see it by printing your computation graph:
```
        self.next_state_computer = function(list(contexts.values()) + states,
                                            next_states + [next_probs])
        from theano.printing import pydotprint
        pydotprint(self.next_state_computer, "next_state.png")
```